main/exec.go

```diff
@@ -25,11 +26,73 @@ func Exec(ctx *log.Context, cmd, workdir string, stdout, stderr io.WriteCloser,
 	defer stdout.Close()
 	defer stderr.Close()
 
+	scriptPath := cmd
+	scriptPathDirectory := filepath.Dir(cmd)
+
 	commandArgs, err := SetEnvironmentVariables(cfg)
-	cmd = cmd + commandArgs // Add command args if any
+	// Add command args if any. Unnamed arguments go in 'commandArgs'. Named arguments are set as environment variables so the'd be available within the script.
+	cmd = cmd + commandArgs
 
 	//executionMessage := ""   // TODO: return
 	exitCode := 0 // TODO: return exit code and execution state
+
```

Comment:
Is this section of the code running as runas user? If not the created script file will have to have its owner changed to the runas user

Reply:
root is owner of /var/lib/waagent/run-command-handler/download/<RunCommandName>/<sequenceNumber>/runAsScript.sh. However, this file contains script passed by user that will be run as RunAsUser on the 4th line. First 3 lines in the file copies the script under /home/<runAsUser>/waagent/run-command-handler-runas/download/<RunCommandName>/<sequenceNumber>


Particulary this bit:
sudo -S -u %s %s", cfg.protectedSettings.RunAsPassword, cfg.publicSettings.RunAsUser, runAsScriptFilePath+commandArgs
(line 4 in /var/lib/waagent/run-command-handler/download/<RunCommandName>/<sequenceNumber>/runAsScript.sh)

If you refer the screenshot of example in PR description, 
id command (part of script) returns information about the runAsUser "vickyl" that I added using "sudo addusers". "vivl" is the admin user that was used to create the VM.


FYI. sample /var/lib/waagent/run-command-handler/download/<RunCommandName>/<sequenceNumber>/runAsScript.sh looks like below:
====
Line 1: mkdir -p -m a=rwx /home/<runAsUserName>/waagent/run-command-handler-runas/download/<RunCommandName>/<sequenceNumber>
Line 2: cp /var/lib/waagent/run-command-handler/download/<RunCommandName>/<sequenceNumber>/script.sh /home/<runAsUserName>/waagent/run-command-handler-runas/download/<RunCommandName>/<sequenceNumber>
Line 3: chmod 555 /home/<runAsUserName>/waagent/run-command-handler-runas/download/<RunCommandName>/<sequenceNumber>/script.sh 
Line 4: echo <runAsPassword> | sudo -S -u <runAsUserName> /home/<runAsUserName>/waagent/run-command-handler-runas/download/<RunCommandName>/<sequenceNumber>/script.sh value1 value2
====

In line 3, chmod provides runAsUser permission to execute the script 
 /home/<runAsUserName>/waagent/run-command-handler-runas/download/<RunCommandName>/<sequenceNumber>/script.sh

Reply:
Is the runAsUser set as the owner of the dir /home//waagent/run-command-handler-runas/download, and script.sh in it?

Reply:
root is still owner of below directory. However, runAsScript.sh provides access to other users to read and execute.
root@VivL1:/home/vickyl/waagent/run-command-handler-runas# ls -lrt
total 4
drwxr-xr-x 3 root root 4096 Oct 25 21:46 download

Should we provide write access to other users as well ?  (runAsUser would be under other user if root is the owner, I believe)

Reply:
addressed comment. Now they read as:
root@VivL1:/home/vickyl/waagent/run-command-handler-runas/download/RA_1025_13# ls -lrt
total 4
drwxr-xr-x 2 vickyl root 4096 Oct 25 23:04 0

root@VivL1:/home/vickyl/waagent/run-command-handler-runas/download/RA_1025_13/0# ls -lrt
total 4
-r-x------ 1 vickyl root 19 Oct 25 23:04 script.sh

user.lookup() does not return any error for some reason when user does not exists (error expected according to docs). Looks like there is a bug in Go lang. I have created a bug to track it [Bug 16003130](https://msazure.visualstudio.com/One/_workitems/edit/16003130): [RCv2 Linux] user.Lookup() does not return any error if the user does not exist.
I'll look into a workaround later for this. It does throw appropriate error right now, like "uknown user"

Reply:
Added a small workaround for RunAs user not found error. Not pretty, but does the job.